### PR TITLE
Improve `Function.list_schedules()`

### PR DESCRIPTION
**Pull request by ddonukis**
## Description
Looking at the ticket: https://cognitedata.atlassian.net/jira/software/projects/DOGE/boards/871?selectedIssue=DOGE-62

I agree that the implementation looks unusual but I can't reproduce a duplication issue.

## Checklist:
- [x] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [x] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [x] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/client/data_classes/functions.py [hunk 1]
```diff
@@ -134,15 +134,7 @@ def list_schedules(self, limit: int | None = DEFAULT_LIMIT_READ) -> FunctionSche
...
-        schedules_by_external_id = self._cognite_client.functions.schedules.list(
-            function_external_id=self.external_id, limit=limit
-        )
-        schedules_by_id = self._cognite_client.functions.schedules.list(function_id=self.id, limit=limit)
-
-        if is_unlimited(limit):
-            limit = self._cognite_client.functions.schedules._LIST_LIMIT_CEILING
-
-        return (schedules_by_external_id + schedules_by_id)[:limit]
+        return self._cognite_client.functions.schedules.list(function_id=self.id, limit=limit)
```
Comment (ddonukis) [inline cognite/client/data_classes/functions.py:137]: According to the function docs `id` is required, so I don't see why we'd need to try getting the schedules by both external id and id, so simplified it. Although the old logic seemed to work just fine.

Comment (haakonvt) [reply cognite/client/data_classes/functions.py:136]: I swear I got duplicates returned the other day... but I'm (also) unable to reproduce it now... I should have taken a screenshot or something... 

Either way, this is a great improvement! 🥇

File: tests/tests_integration/test_api/test_functions.py [hunk 2]
```diff
@@ -16,3 +16,13 @@ def test_retrieve_unknown_raises_error(self, cognite_client: CogniteClient):
     def test_retrieve_unknown_ignore_unknowns(self, cognite_client: CogniteClient):
         res = cognite_client.functions.retrieve_multiple(external_ids=["this does not exist"], ignore_unknown_ids=True)
         assert len(res) == 0
+
+    def test_function_list_schedules_unlimited(self, cognite_client: CogniteClient):
+        expected_unique_schedules = 100
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_functions.py:21]: Let's pick a lower number (integration tests put quite a lot of pressure on CDF already). Maybe 5?

File: cognite/client/data_classes/functions.py (hunk 1 above)
Comment (haakonvt) [reply cognite/client/data_classes/functions.py:136]: Puh 😅 😅 😅  managed to recreate the bug now:

image

Comment (haakonvt) [reply cognite/client/data_classes/functions.py:136]: Checking out this branch, I got what I expected:

image

Comment (haakonvt) [reply cognite/client/data_classes/functions.py:136]: Oh, I'm stupid... 🤦  When a function is missing `external_id`, you basically fetch all schedules in the project (on line `137-139`):
```py
schedules_by_external_id = self._cognite_client.functions.schedules.list(
    function_external_id=self.external_id, limit=limit
)

**Comment by haakonvt** [comment]
Now that the bug is identified, let's make a realse of this. Remember version + change log as well 😄

File: cognite/client/data_classes/functions.py (hunk 1 above)
Comment (haakonvt) [reply cognite/client/data_classes/functions.py:136]: fyi @HMEiding @qTipTip

File: tests/tests_integration/test_api/test_functions.py [hunk 3]
```diff
@@ -16,3 +16,13 @@ def test_retrieve_unknown_raises_error(self, cognite_client: CogniteClient):
...
+
+    def test_function_list_schedules_unlimited(self, cognite_client: CogniteClient):
+        expected_unique_schedules = 5
+        # This is an integration test dummy function that purposefully doesn't have an external id.
+        fn = cognite_client.functions.retrieve(id=2495645514618888)
+        schedules = fn.list_schedules(limit=-1)
+
+        assert len(schedules) == expected_unique_schedules
+        assert len({s.id for s in schedules}) == expected_unique_schedules
+        assert len({s.cron_expression for s in schedules}) == expected_unique_schedules
```
Comment (ddonukis) [inline tests/tests_integration/test_api/test_functions.py:28]: This test now catches the duplication error. I created other dummy functions in the project with schedules, with and without external id.

Tbh, I think the behaviour of the schedules/list is not what I'd expect if external_id = None. I though it'd return an empty list 🤷‍♂️

File: tests/tests_unit/test_data_classes/test_functions.py [hunk 4]
```diff
@@ -90,24 +69,6 @@ def test_update_on_deleted_function(self, empty_function):
         empty_function._cognite_client.functions.retrieve.return_value = None
         empty_function.update()
 
-    @pytest.mark.parametrize("limit", [2, None, -1, float("inf")])
-    def test_list_schedules(self, function, function_schedules, limit):
```
Comment (ddonukis) [inline tests/tests_unit/test_data_classes/test_functions.py:94]: This test was literally enforcing the bad logic!

Comment (ddonukis) [reply tests/tests_unit/test_data_classes/test_functions.py:94]: Getting rid of it, not point unit testing a one line wrapper that is already tested in integration.

**Comment by ddonukis** [comment]
All should be good to go 👌

File: tests/tests_integration/test_api/test_functions.py [hunk 5]
```diff
@@ -16,3 +16,13 @@ def test_retrieve_unknown_raises_error(self, cognite_client: CogniteClient):
     def test_retrieve_unknown_ignore_unknowns(self, cognite_client: CogniteClient):
         res = cognite_client.functions.retrieve_multiple(external_ids=["this does not exist"], ignore_unknown_ids=True)
         assert len(res) == 0
+
+    def test_function_list_schedules_unlimited(self, cognite_client: CogniteClient):
+        expected_unique_schedules = 5
+        # This is an integration test dummy function that purposefully doesn't have an external id.
```
Comment (haakonvt) [inline tests/tests_integration/test_api/test_functions.py:22]: Then add an assert for it 👌

File: tests/tests_unit/test_data_classes/test_functions.py (hunk 4 above)
Comment (haakonvt) [reply tests/tests_unit/test_data_classes/test_functions.py:94]: Agreed 👍👍

File: cognite/client/data_classes/functions.py [hunk 6]
```diff
@@ -134,15 +133,7 @@ def list_schedules(self, limit: int | None = DEFAULT_LIMIT_READ) -> FunctionSche
...
```
Comment (haakonvt) [inline cognite/client/data_classes/functions.py:136]: Could add in an if-test on `id` while you’re at it?

**Review by haakonvt — APPROVED** [review]
Good stuff!

**Pull request merged**